File: internal/controller/keystone.go
package controller

import (
	"context"
	"crypto/sha256"
	"crypto/x509"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"time"

	"github.com/kunorikiku/yet-another-overlay-generator/internal/apierr"
	"github.com/kunorikiku/yet-another-overlay-generator/internal/bundlesig"
	"github.com/kunorikiku/yet-another-overlay-generator/internal/trustlist"
)

// bundleSHA256 is the lowercase-hex SHA-256 of a bundle's checksums.sha256 bytes — the
// digest bound into each member of the off-host-signed manifest. checksums.sha256 covers
// install.sh AND every config, so this single digest pins the entire deployed bundle: a
// tampered install.sh changes checksums.sha256, which changes this digest, which the
// breached controller cannot re-sign without the off-host key. It is computed from the
// SAME bytes the agent re-derives (files["checksums.sha256"]).
func bundleSHA256(checksums []byte) string {
	sum := sha256.Sum256(checksums)
	return hex.EncodeToString(sum[:])
}

// memberKey is the comparable identity of a manifest member used by the monotonic-epoch
// rule: the tuple (wg_public_key, bundle_sha256) keyed by node_id. Two manifests carry
// the same membership iff they map the same node_id set to the same tuples.
type memberKey struct {
	wgPublicKey  string
	bundleSHA256 string
}

// manifestMembers decodes a stored manifest's canonical JSON into a node_id -> memberKey
// map for membership comparison.
func manifestMembers(trustListJSON []byte) (map[string]memberKey, error) {
	var tl trustlist.TrustList
	if err := json.Unmarshal(trustListJSON, &tl); err != nil {
		return nil, fmt.Errorf("controller: parsing stored manifest: %w", err)
	}
	out := make(map[string]memberKey, len(tl.Members))
	for _, m := range tl.Members {
		out[m.NodeID] = memberKey{wgPublicKey: m.WGPublicKey, bundleSHA256: m.BundleSHA256}
	}
	return out, nil
}

// sameMembership reports whether two node_id -> memberKey maps are equal (same node set,
// same tuple per node). It is the freshness test the monotonic-epoch rule uses to decide
// whether to REUSE the stored epoch (identical signed content) or BUMP it.
func sameMembership(a, b map[string]memberKey) bool {
	if len(a) != len(b) {
		return false
	}
	for id, ka := range a {
		kb, ok := b[id]
		if !ok || ka != kb {
			return false
		}
	}
	return true
}

// enforceSigningAnchor reconciles the configured bundle signing key (YAOG_BUNDLE_SIGNING_KEY,
// resolved here exactly as artifacts.Export will) against the tenant's persisted SigningAnchor, so
// a redeploy that dropped or swapped the key is caught BEFORE any bundle is produced:
//
//   - no anchor + key present → pin it (trust-on-first-use), then sign as usual
//   - no anchor + no key      → never-signed fleet, allowed (back-compat, hash-only bundles)
//   - anchor + same key       → normal signed stage
//   - anchor + NO key         → CodeSigningKeyMissing (refuse: would silently downgrade to unsigned)
//   - anchor + different key  → CodeSigningKeyMismatch, unless YAOG_BUNDLE_SIGNING_KEY_ROTATE re-pins
//
// The two refusal cases return a coded *apierr.Error so the operator gets a precise reason on stage.
func enforceSigningAnchor(ctx context.Context, store Store, t TenantID, now time.Time) error {
	signer, err := bundlesig.LoadConfigSignerFromEnv()
	if err != nil {
		// A set-but-unreadable/unparsable key already fails the export closed; surface it here too
		// so a half-configured signer never slips past the anchor reconciliation.
		return fmt.Errorf("controller: loading bundle signing key: %w", err)
	}
	var configuredPub string
	if signer != nil {
		configuredPub = string(signer.PublicKeyPEM())
	}

	anchor, err := store.GetSigningAnchor(ctx, t)
	switch {
	case errors.Is(err, ErrNotFound):
		if configuredPub == "" {
			return nil // never-signed fleet: nothing to pin, nothing to enforce
		}
		if err := store.PutSigningAnchor(ctx, t, SigningAnchor{PubKeyPEM: configuredPub}); err != nil {
			return err
		}
		// Trust-on-first-use: this re-points which key the fleet is signed under, so audit it
		// (best-effort, like the stage audits) — a trust transition must be attributable.
		appendStageAudit(ctx, store, t, now, "signing-anchor-pin", "")
		return nil
	case err != nil:
		return fmt.Errorf("controller: loading signing anchor: %w", err)
	}

	switch {
	case configuredPub == "":
		return apierr.New(apierr.CodeSigningKeyMissing) // pinned-but-absent → refuse
	case configuredPub == anchor.PubKeyPEM:
		return nil // same key as pinned — normal signed stage
	case bundlesig.RotateRequested():
		if err := store.PutSigningAnchor(ctx, t, SigningAnchor{PubKeyPEM: configuredPub}); err != nil {
			return err
		}
		// Explicit rotation (YAOG_BUNDLE_SIGNING_KEY_ROTATE) — audit the re-pin so a key change is
		// attributable in the hash-chained log, not indistinguishable from a routine stage.
		appendStageAudit(ctx, store, t, now, "signing-anchor-rotate", "")
		return nil
	default:
		return apierr.New(apierr.CodeSigningKeyMismatch) // configured key != pinned → refuse
	}
}

// stageManifest assembles the off-host-signable membership manifest from the staged
// nodes — each Member is {NodeID, WGPublicKey, BundleSHA256} — and stores it as the
// staged, UNSIGNED manifest (StoredTrustList.TrustListJSON = Canonical(manifest),
// SignatureJSON empty, Epoch set by the monotonic rule). The members are exactly the
// nodes that were rendered this stage (only they carry a bundle digest); their WG public
// keys come from the registry value stamped on the subgraph.
//
// Monotonic epoch (anti-rollback): reuse the prior stored manifest's epoch iff its
// membership (node_id -> {wg key, bundle digest}) is byte-for-byte the same; otherwise
// prior-epoch+1, or 0 when no manifest has ever been stored. Because BundleSHA256 is now
// part of the membership tuple, ANY change to a node's install.sh/config (which changes
// its bundle digest) advances the epoch, so a node's anti-rollback floor admits the fresh
// deploy and rejects a stale one.
func stageManifest(ctx context.Context, store Store, t TenantID, digests, pubKeys map[string]string) error {
	members := make([]trustlist.Member, 0, len(digests))
	for nodeID, dig := range digests {
		members = append(members, trustlist.Member{
			NodeID:       nodeID,
			WGPublicKey:  pubKeys[nodeID],
			BundleSHA256: dig,
		})
	}

	newMembers := make(map[string]memberKey, len(members))
	for _, m := range members {
		newMembers[m.NodeID] = memberKey{wgPublicKey: m.WGPublicKey, bundleSHA256: m.BundleSHA256}
	}

	// Monotonic epoch relative to the prior STAGED manifest (GetCurrentSignedTrustList) — NOT the
	// served slot. Chaining off the staging history keeps in-flight re-stages monotonic; because the
	// served epoch is always a subsequence of staged epochs (served only ever takes a value that was
	// once staged, at promote), anti-rollback on the served path is preserved. Do not "simplify" this
	// to read the served slot — that would lose monotonicity across un-promoted re-stages.
	var epoch int64
	if stored, err := store.GetCurrentSignedTrustList(ctx, t); err == nil {
		priorMembers, perr := manifestMembers(stored.TrustListJSON)
		if perr != nil {
			return perr
		}
		if sameMembership(newMembers, priorMembers) {
			epoch = stored.Epoch
		} else {
			epoch = stored.Epoch + 1
		}
	} else if !errors.Is(err, ErrNotFound) {
		return fmt.Errorf("controller: loading prior manifest for epoch: %w", err)
	}

	manifest := trustlist.TrustList{
		SchemaVersion: 1,
		Tenant:        string(t),
		Epoch:         epoch,
		Members:       members,
	}
	canonical, err := trustlist.Canonical(manifest)
	if err != nil {
		return fmt.Errorf("controller: canonicalizing staged manifest: %w", err)
	}

	// Store the staged manifest with an EMPTY signature: staging never requires a
	// signature. The operator signs it off-host (GET /trustlist → POST
	// /trustlist-signature, which sets SignatureJSON), and PromoteStaged refuses until
	// that signature exists, matches these bytes, and verifies.
	if err := store.PutSignedTrustList(ctx, t, StoredTrustList{
		TrustListJSON: canonical,
		SignatureJSON: nil,
		Epoch:         epoch,
	}); err != nil {
		return fmt.Errorf("controller: storing staged manifest: %w", err)
	}
	return nil
}

// pinFromOperatorCredential builds the trustlist.PinnedCredential the verifier checks
// against from a stored OperatorCredential, parsing the PEM by the credential's
// algorithm. It mirrors the HTTP layer's pinFromCredential so the promote-gate verifies
// with exactly the anchor a node would use.
func pinFromOperatorCredential(c OperatorCredential) (trustlist.PinnedCredential, error) {
	pin := trustlist.PinnedCredential{
		Alg:          trustlist.Alg(c.Alg),
		CredentialID: c.CredentialID,
		RPID:         c.RPID,
		Origin:       c.Origin,
	}
	switch trustlist.Alg(c.Alg) {
	case trustlist.AlgEd25519, trustlist.AlgWebAuthnEdDSA:
		pub, err := trustlist.ParseEd25519PinPEM([]byte(c.PublicKeyPEM))
		if err != nil {
			return trustlist.PinnedCredential{}, err
		}
		pin.Ed25519Pub = pub
	case trustlist.AlgWebAuthnES256:
		pub, err := trustlist.ParseES256Pin([]byte(c.PublicKeyPEM))
		if err != nil {
			return trustlist.PinnedCredential{}, err
		}
		pin.ES256Pub = pub
	default:
		return trustlist.PinnedCredential{}, fmt.Errorf("controller: unsupported operator credential algorithm %q", c.Alg)
	}
	return pin, nil
}

// KeystoneFingerprint is the stable hex SHA-256 of a pinned operator credential's PUBLIC
// key, hashed over its CANONICAL x509 PKIX DER (never the PEM string) so a benign
// re-encode — a trailing newline, different line wrapping — of the SAME key yields the
// SAME fingerprint. It is the single identity used to tell a keystone ROTATION (a genuinely
// different key) apart from an idempotent re-pin, and is surfaced to the operator panel as
// a short, comparable credential identity. It is NON-SECRET (a public-key digest). The
// alg→parse dispatch is reused from pinFromOperatorCredential so it lives in one place.
func KeystoneFingerprint(c OperatorCredential) (string, error) {
	pin, err := pinFromOperatorCredential(c)
	if err != nil {
		return "", err
	}
	var pub any
	switch {
	case pin.Ed25519Pub != nil:
		pub = pin.Ed25519Pub
	case pin.ES256Pub != nil:
		pub = pin.ES256Pub
	default:
		return "", fmt.Errorf("controller: operator credential has no public key to fingerprint")
	}
	der, err := x509.MarshalPKIXPublicKey(pub)
	if err != nil {
		return "", fmt.Errorf("controller: marshalling operator credential public key: %w", err)
	}
	sum := sha256.Sum256(der)
	return hex.EncodeToString(sum[:]), nil
}

// SameKeystoneCredential reports whether two operator credentials are the SAME trust anchor:
// the same public key (by fingerprint), the same algorithm, and — for the WebAuthn algorithms
// whose assertion binds them — the same relying-party ID, origin, and credential id. Any
// difference is a ROTATION (a new key, or a rebinding that changes what trustlist.Verify
// accepts), which the handler refuses without an explicit ack. RPID/Origin/CredentialID all
// feed trustlist.Verify (rpIdHash + the fail-closed origin check + allowCredentials), so a
// change in any of them is a real trust shift, not cosmetic. A fingerprint error (an unparsable
// PEM) is returned to the caller rather than masked as "same".
func SameKeystoneCredential(a, b OperatorCredential) (bool, error) {
	if a.Alg != b.Alg {
		return false, nil
	}
	fpA, err := KeystoneFingerprint(a)
	if err != nil {
		return false, err
	}
	fpB, err := KeystoneFingerprint(b)
	if err != nil {
		return false, err
	}
	if fpA != fpB {
		return false, nil
	}
	// The WebAuthn assertion binds rpid + origin + credential id (all consumed by
	// trustlist.Verify), so a change in ANY of them changes what verifies even with the same
	// key; raw ed25519 ignores all three (empty on both sides).
	switch trustlist.Alg(a.Alg) {
	case trustlist.AlgWebAuthnES256, trustlist.AlgWebAuthnEdDSA:
		return a.RPID == b.RPID && a.Origin == b.Origin && a.CredentialID == b.CredentialID, nil
	default:
		return true, nil
	}
}

// verifyStoredAgainstPin is the single trust-critical parse-and-verify of a stored signed
// manifest against a pinned credential, shared by the promote gate (PromoteStaged) and the
// redeploy-required signal (KeystoneRedeployRequired) so the two can never drift. It SEPARATES
// the two failure kinds the callers must treat differently:
//
//   - parseErr: a corrupt stored record or an unparsable pinned credential — a real fault.
//   - verifyErr: a well-formed signature that does NOT verify against the pin — for the redeploy
//     signal this is precisely the "rotated away from the served key" condition, not a fault.
//
// Exactly one is non-nil on failure; both are nil on success.
func verifyStoredAgainstPin(stored StoredTrustList, cred OperatorCredential) (parseErr, verifyErr error) {
	var manifest trustlist.TrustList
	if err := json.Unmarshal(stored.TrustListJSON, &manifest); err != nil {
		return fmt.Errorf("controller: parsing stored manifest: %w", err), nil
	}
	var signed trustlist.SignedTrustList
	if err := json.Unmarshal(stored.SignatureJSON, &signed); err != nil {
		return fmt.Errorf("controller: parsing stored signature: %w", err), nil
	}
	pin, err := pinFromOperatorCredential(cred)
	if err != nil {
		return err, nil
	}
	return nil, trustlist.Verify(manifest, signed, pin)
}

// KeystoneRedeployRequired reports whether the SERVED signed membership trust-list (read from the
// served slot via GetServedTrustList — exactly what /config hands the fleet) carries a signature
// that no longer verifies against the pinned credential `cred` — i.e. the keystone was rotated but
// no fresh deploy has been signed + promoted under the new key yet, so every (correctly
// re-provisioned) node would refuse the served bundle. It is the single source of the operator-
// facing "redeploy required" signal.
//
// It is deliberately CONSERVATIVE: it returns false (not "required") when nothing has been promoted
// (served slot empty — a stage→sign→promote may be mid-flight in the STAGED slot, which this
// function does not read) or when the served signature still verifies — so it fires ONLY for a
// genuine rotated-but-not-redeployed fleet, never mid-deploy. A parse/pin error (a corrupt stored
// record or an unparsable pinned credential) is surfaced to the caller, never masked as "not
// required".
func KeystoneRedeployRequired(ctx context.Context, store Store, t TenantID, cred OperatorCredential) (bool, error) {
	stored, err := store.GetServedTrustList(ctx, t)
	if errors.Is(err, ErrNotFound) {
		return false, nil // nothing promoted/served yet — not a rotation
	}
	if err != nil {
		return false, fmt.Errorf("controller: loading served trust-list for redeploy check: %w", err)
	}
	if len(stored.SignatureJSON) == 0 {
		// Defensive only: the served slot is signed-only by construction (both PromoteStaged writers
		// gate the served copy on a non-empty signature, and the keystone-ON promote gate refuses an
		// unsigned promote). Treat an impossible unsigned served manifest as not-required.
		return false, nil
	}
	parseErr, verifyErr := verifyStoredAgainstPin(stored, cred)
	if parseErr != nil {
		return false, parseErr // a real fault — never silently "not required"
	}
	// verifyErr != nil is the rotated-away-from-the-served-key signal: redeploy required.
	return verifyErr != nil, nil
}

File: frontend/src/api/controller/keystone.ts
// Keystone (off-host operator signing) client routes + wire types.
//
// These mirror internal/trustlist/types.go (SignedTrustList) and the keystone routes in
// internal/api/handler_keystone.go. The byte-level contract — every base64url field is RFC4648
// url-alphabet WITHOUT padding (Go base64.RawURLEncoding), the challenge binding, and the rpid
// binding — lives in ../../lib/webauthn.ts, which is the single place that builds these structs.

import { request, postJSON, ControllerError, type ControllerConfig } from './transport';

// WebAuthnAlg is the keystone signing algorithm. Only ES256 and EdDSA WebAuthn
// assertions are accepted by the node verifier; RS256 etc. are rejected.
export type WebAuthnAlg = 'webauthn-es256' | 'webauthn-eddsa';

// Wire constants for the two accepted algorithms (kept here so both the client
// and the webauthn helper import the literal from one place).
export const AlgWebAuthnES256 = 'webauthn-es256' as const;
export const AlgWebAuthnEdDSA = 'webauthn-eddsa' as const;

// SignedTrustList is the historical wire name for a WebAuthn assertion. It normally carries the
// detached signature over a deploy's canonical manifest, and the same shape carries random-
// challenge login and enrollment assertions. Field names + base64url encodings match Go exactly.
export interface SignedTrustList {
  alg: WebAuthnAlg;
  credential_id: string; // base64url(rawId)
  public_key: string; // pinned PKIX PEM (audit only; node verifies the PINNED key)
  signature: string; // base64url(response.signature) — ES256 is ASN.1 DER
  authenticator_data: string; // base64url(response.authenticatorData)
  client_data_json: string; // base64url(response.clientDataJSON)
}

// trustListResponseJSON shape from GET /trustlist: the canonical manifest bytes
// (STANDARD base64) to be signed, plus the membership epoch they carry.
interface TrustListResponseJSON {
  trustlist_json: string; // standard base64 of the canonical manifest bytes
  epoch: number;
}

// The panel-facing GET /trustlist result: trustlistJson is STANDARD base64 (the
// caller base64-decodes it to recover the canonical bytes whose SHA-256 is the
// WebAuthn challenge). A null return means the keystone is OFF for the tenant
// (404 = no operator credential pinned / nothing staged to sign).
export interface TrustListToSign {
  trustlistJson: string;
  epoch: number;
}

// operatorCredentialRequestJSON shape for POST /operator-credential: the pinned
// off-host signing credential. public_key_pem is the PKIX "PUBLIC KEY" PEM; rpid
// MUST equal the rp.id used at create() time (the verifier binds SHA256(rpid) to the
// assertion rpIdHash). New/rotated browser credentials also carry enrollmentProof.
export interface OperatorCredentialBody {
  alg: WebAuthnAlg;
  credentialId: string; // base64url(rawId)
  publicKeyPEM: string; // PKIX "PUBLIC KEY" PEM
  rpId: string; // location.hostname
  origin: string; // location.origin
  enrollmentProof: SignedTrustList; // candidate's UV assertion over a one-use server challenge
  // rotate ACKNOWLEDGES that this pin REPLACES a different already-pinned credential (which
  // strands every node until re-provisioned + redeployed). The controller refuses a changed
  // credential without it (409 keystone_rotation_requires_ack). Ignored on a first/idempotent pin.
  rotate?: boolean;
}

// OperatorCredentialPinResult is the POST /operator-credential result: rotated true only when the
// pin REPLACED a different credential; unchanged true on an idempotent re-pin; redeployRequired
// true when (after a rotation) the served fleet is still signed under the old key.
export interface OperatorCredentialPinResult {
  ok: boolean;
  rotated: boolean;
  unchanged: boolean;
  redeployRequired: boolean;
}

// OperatorCredentialStatus is the SERVER-authoritative keystone status (GET /operator-credential):
// the panel reflects THIS, never a browser-local cache, so a cleared browser can never falsely
// read "Not enrolled". It carries only non-secret public identifiers. publicKeyPEM is the
// credential's PUBLIC PEM (audit-only, already baked into every node bundle): a cleared/fresh
// browser recovers it — with credentialId/alg/rpId — to re-prompt the authenticator WITHOUT
// re-pinning (plan-3 signing-handle auto-recovery). Only public descriptor material is returned;
// YAOG does not receive plaintext private-key material or establish provider non-exportability.
export interface OperatorCredentialStatus {
  pinned: boolean;
  alg: string;
  credentialId: string;
  rpId: string;
  origin: string;
  fingerprint: string;
  publicKeyPEM: string;
  redeployRequired: boolean;
}

// getTrustlist fetches the STAGED membership manifest the operator must sign
// (operator-only). It returns the canonical bytes as STANDARD base64 plus the
// epoch — the panel base64-decodes trustlistJson and signs SHA-256 of those bytes.
//
// A 404 means the keystone is OFF (no operator credential pinned, or nothing
// staged): the handler returns 404 from GET /trustlist when there is no staged
// manifest, and the keystone is only ON once a credential is pinned. We map 404
// to null so deploy() can promote directly (today's behavior) when the keystone
// is off, and only run the signing ceremony when a manifest comes back.
export async function getTrustlist(cfg: ControllerConfig): Promise<TrustListToSign | null> {
  try {
    const res = await request(cfg, 'trustlist', { method: 'GET' });
    const data = (await res.json()) as TrustListResponseJSON;
    return { trustlistJson: data.trustlist_json, epoch: data.epoch };
  } catch (err) {
    // request() throws ControllerError on non-2xx; 404 = keystone OFF (no operator
    // credential pinned, or nothing staged) → null so deploy() promotes directly.
    // Match on the typed status, not the message. (Previously this used a raw fetch
    // WITHOUT credentials:'include', so a refreshed cookie-only operator session 401'd
    // and could not keystone-sign on Deploy on a keystone-ON tenant — F1.)
    if (err instanceof ControllerError && err.status === 404) {
      return null;
    }
    throw err;
  }
}

// postTrustlistSignature submits the operator's off-host signature over the
// staged manifest (operator-only). trustlistJson is the STANDARD base64 of the
// exact bytes signed (server-side substitution guard) and signed is the
// SignedTrustList assembled by ../../lib/webauthn.ts. A non-2xx (e.g. 400 verify
// failure, 409 manifest changed, 412 no credential pinned) throws as usual.
export async function postTrustlistSignature(
  cfg: ControllerConfig,
  body: { trustlistJson: string; signed: SignedTrustList },
): Promise<void> {
  await postJSON(
    cfg,
    'trustlist-signature',
    JSON.stringify({ trustlist_json: body.trustlistJson, signed: body.signed }),
  );
}

// postOperatorCredential pins the off-host operator signing credential, turning
// the keystone ON for the tenant (operator-only). The body's PEM must parse for
// the declared alg server-side (a malformed pin is a 400). rpid/origin carry the
// WebAuthn relying-party binding the node enforces.
export async function postOperatorCredential(
  cfg: ControllerConfig,
  body: OperatorCredentialBody,
): Promise<OperatorCredentialPinResult> {
  const res = await postJSON(
    cfg,
    'operator-credential',
    JSON.stringify({
      alg: body.alg,
      credential_id: body.credentialId,
      public_key_pem: body.publicKeyPEM,
      rpid: body.rpId,
      origin: body.origin,
      enrollment_proof: body.enrollmentProof,
      rotate: body.rotate ?? false,
    }),
  );
  const d = (await res.json()) as {
    ok?: boolean;
    rotated?: boolean;
    unchanged?: boolean;
    redeploy_required?: boolean;
  };
  return {
    ok: d.ok ?? false,
    rotated: d.rotated ?? false,
    unchanged: d.unchanged ?? false,
    redeployRequired: d.redeploy_required ?? false,
  };
}

// getOperatorCredentialStatus reads the SERVER-authoritative keystone status (GET
// /operator-credential, operator-only). It is the source of truth for the panel's "enrolled"
// display — a browser-local cache must never decide it (a cleared browser would falsely read
// "Not enrolled" and invite a fleet-stranding re-pin).
export async function getOperatorCredentialStatus(
  cfg: ControllerConfig,
): Promise<OperatorCredentialStatus> {
  const res = await request(cfg, 'operator-credential', { method: 'GET' });
  const d = (await res.json()) as {
    pinned?: boolean;
    alg?: string;
    credential_id?: string;
    rpid?: string;
    origin?: string;
    fingerprint?: string;
    public_key_pem?: string;
    redeploy_required?: boolean;
  };
  // Security-sensitive tri-state boundary: an absent/malformed `pinned` value is UNKNOWN, not
  // keystone-off. Callers deliberately retain their prior/null status on this error, so the panel
  // cannot expose the dangerous no-keystone workflow without an explicit false from the server.
  if (typeof d.pinned !== 'boolean') {
    throw new Error('operator credential status response is missing boolean pinned');
  }
  return {
    pinned: d.pinned,
    alg: d.alg ?? '',
    credentialId: d.credential_id ?? '',
    rpId: d.rpid ?? '',
    origin: d.origin ?? '',
    fingerprint: d.fingerprint ?? '',
    publicKeyPEM: d.public_key_pem ?? '',
    redeployRequired: d.redeploy_required ?? false,
  };
}
